<a href="https://colab.research.google.com/github/merblaine/Week4/blob/main/Copy_of_Week_3_Homework.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Week 3 Homework

(Be sure to copy to drive first)

For this homework we are going to apply some text analysis stuff to a collection of PDF file that I'll ask you to put together.

In [ ]:
#Install / load our libraries and connect to Google Drive

!pip install pypdf

#Colab has a special library just for working with files on google drive
from google.colab import drive


#This is a new library we'll use to extract text from PDFs
from pypdf import PdfReader


import textblob
import pandas as pd
import numpy as np
import glob

#Some extra libraries we'll need for text analysis
import nltk
nltk.download('punkt')
nltk.download('brown')
nltk.download('punkt_tab')

#Connect to Gdrive
drive.mount('/content/gdrive')

print("Libraries and Drive Ready!")

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 310.5/310.5 kB 9.2 MB/s eta 0:00:00


[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.
[nltk_data] Downloading package brown to /root/nltk_data...
[nltk_data]   Unzipping corpora/brown.zip.
[nltk_data] Downloading package punkt_tab to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt_tab.zip.


Mounted at /content/gdrive
Libraries and Drive Ready!


**H1**


Part of the work for this week is going to be finding a collection of 6 or so PDFs. These PDFS ideally should be on the same thing (journal articles on the same topic). Additionally these PDFs whould be Open Access, or the very least obtained lawfully * ***ahem*** *.

Copy these 6 PDFs into the `LibraryJuicePython` folder on Google Drive. We'll be peforming some text analysis on these files.

In [ ]:
#Run this cell to see the files in your directory
for file in glob.glob("/content/gdrive/MyDrive/LibraryJuicePython/*.pdf"):
  print(file)

/content/gdrive/MyDrive/LibraryJuicePython/03_article_00.pdf
/content/gdrive/MyDrive/LibraryJuicePython/03_article_01.pdf
/content/gdrive/MyDrive/LibraryJuicePython/03_article_05.pdf
/content/gdrive/MyDrive/LibraryJuicePython/03_article_04.pdf
/content/gdrive/MyDrive/LibraryJuicePython/03_article_03.pdf
/content/gdrive/MyDrive/LibraryJuicePython/03_article_02.pdf


# Exracting the text from the PDFs

We'll use a Python Library to create a text file of the contents of the PDF file.

In [ ]:

for file in glob.glob("/content/gdrive/MyDrive/LibraryJuicePython/*.pdf"):
  filename = file.split("/")[-1]
  print("Extracting text for... ",filename)

  text = ""
  #This bit is new!
  reader = PdfReader(file)
  for page in reader.pages:
    text += page.extract_text()

  #Output the string variable into a text file
  output_text_file = filename+".txt"
  with open(output_text_file, "w") as text_file:
    text_file.write(text)

Extracting text for...  03_article_00.pdf
Extracting text for...  03_article_01.pdf
Extracting text for...  03_article_05.pdf
Extracting text for...  03_article_04.pdf
Extracting text for...  03_article_03.pdf
Extracting text for...  03_article_02.pdf


# Sentiment of our PDFs

Let's just print our sentiment scores to the screen for now.

In [ ]:
for file in glob.glob("*.txt"):
  print("Sentiment for ",file)
  with open(file,"r") as text_file:
    text = text_file.read()
    blob = textblob.TextBlob(text)
    print(blob.sentiment)
    print("---")

Sentiment for  03_article_00.pdf.txt
Sentiment(polarity=0.025863681107867145, subjectivity=0.3559742729277617)
---
Sentiment for  03_article_04.pdf.txt
Sentiment(polarity=0.0798492077182779, subjectivity=0.3877951814156751)
---
Sentiment for  03_article_01.pdf.txt
Sentiment(polarity=0.03610700266544426, subjectivity=0.3374564327336561)
---
Sentiment for  03_article_05.pdf.txt
Sentiment(polarity=0.0650304082722413, subjectivity=0.35300937803206406)
---
Sentiment for  03_article_03.pdf.txt
Sentiment(polarity=0.020321995178224285, subjectivity=0.36302257599031834)
---
Sentiment for  03_article_02.pdf.txt
Sentiment(polarity=0.10515704322028421, subjectivity=0.3962403253312338)
---


# Noun Phrases of our PDFs

We'll just do the top 3 keywords of each file and display it to the screen

In [ ]:
for file in glob.glob("*.txt"):
  print("Top 3 Phrases for ",file)
  with open(file,"r") as text_file:
    text = text_file.read()

  blob = textblob.TextBlob(text)
  nphrases = dict()

  for np in blob.noun_phrases:
    if np in nphrases:
         nphrases[np] += 1
    else:
         nphrases[np] = 1

  for np in sorted(nphrases, key=nphrases.get, reverse=True)[0:3]:
      print(np, nphrases[np])

  print("---")

Top 3 Phrases for  03_article_00.pdf.txt
finland 45
k€ars€am€aki 35
turku 28
---
Top 3 Phrases for  03_article_04.pdf.txt
berenice 60
egypt 18
figure 18
---
Top 3 Phrases for  03_article_01.pdf.txt
sv 51
verona 37
italy 35
---
Top 3 Phrases for  03_article_05.pdf.txt
australia 108
plos one 38
spit 38
---
Top 3 Phrases for  03_article_03.pdf.txt
lithuania 119
ad 112
viking 35
---
Top 3 Phrases for  03_article_02.pdf.txt
driesch 157
kratochvil 84
boessneck 79
---



# Dataframe of results

Let's put our findings into a dataframe and save them to our drive. (We will use these PDFs again in our Week 4 work)

In [ ]:

data_set = []

for file in glob.glob("*.txt"):

  pdf_detail = []

  with open(file,"r") as text_file:
    text = text_file.read()
  blob = textblob.TextBlob(text)

  print("working on ",file)
  pdf_detail.append(file)
  pdf_detail.append(blob.sentiment.polarity)
  pdf_detail.append(blob.sentiment.subjectivity)


  nphrases = dict()
  for np in blob.noun_phrases:
    if np in nphrases:
      nphrases[np] += 1
    else:
      nphrases[np] = 1

  #This is a weird way to find the top three keywords
  #This is an example of 'hacky' code

  #we loop through the first entry [0:1] in the sorted keyword list
  #giving us keyword one
  for np in sorted(nphrases, key=nphrases.get, reverse=True)[0:1]:
      keyword_one = np
      pdf_detail.append(keyword_one)

  #we loop through the first entry [1:2] in the sorted keyword list
  #giving us keyword two
  for np in sorted(nphrases, key=nphrases.get, reverse=True)[1:2]:
      keyword_two = np
      pdf_detail.append(keyword_two)

  #we loop through the first entry [2:3] in the sorted keyword list
  #giving us keyword two
  for np in sorted(nphrases, key=nphrases.get, reverse=True)[2:3]:
      keyword_three = np
      pdf_detail.append(keyword_three)

  data_set.append(pdf_detail)
  print(pdf_detail)
  print("---")


working on  03_article_00.pdf.txt
['03_article_00.pdf.txt', 0.025863681107867145, 0.3559742729277617, 'finland', 'k€ars€am€aki', 'turku']
---
working on  03_article_04.pdf.txt
['03_article_04.pdf.txt', 0.0798492077182779, 0.3877951814156751, 'berenice', 'egypt', 'figure']
---
working on  03_article_01.pdf.txt
['03_article_01.pdf.txt', 0.03610700266544426, 0.3374564327336561, 'sv', 'verona', 'italy']
---
working on  03_article_05.pdf.txt
['03_article_05.pdf.txt', 0.0650304082722413, 0.35300937803206406, 'australia', 'plos one', 'spit']
---
working on  03_article_03.pdf.txt
['03_article_03.pdf.txt', 0.020321995178224285, 0.36302257599031834, 'lithuania', 'ad', 'viking']
---
working on  03_article_02.pdf.txt
['03_article_02.pdf.txt', 0.10515704322028421, 0.3962403253312338, 'driesch', 'kratochvil', 'boessneck']
---


In [ ]:
#We now have a list of lists
#This sorta looks like JSON doesn't it?
data_set

[['03_article_00.pdf.txt',
  0.025863681107867145,
  0.3559742729277617,
  'finland',
  'k€ars€am€aki',
  'turku'],
 ['03_article_04.pdf.txt',
  0.0798492077182779,
  0.3877951814156751,
  'berenice',
  'egypt',
  'figure'],
 ['03_article_01.pdf.txt',
  0.03610700266544426,
  0.3374564327336561,
  'sv',
  'verona',
  'italy'],
 ['03_article_05.pdf.txt',
  0.0650304082722413,
  0.35300937803206406,
  'australia',
  'plos one',
  'spit'],
 ['03_article_03.pdf.txt',
  0.020321995178224285,
  0.36302257599031834,
  'lithuania',
  'ad',
  'viking'],
 ['03_article_02.pdf.txt',
  0.10515704322028421,
  0.3962403253312338,
  'driesch',
  'kratochvil',
  'boessneck']]

In [ ]:
#Let's turn this list of lists into a pandas dataframe
pdf_dataframe = pd.DataFrame(data_set)
pdf_dataframe

0         1         2          3             4  \
0  03_article_00.pdf.txt  0.025864  0.355974    finland  k€ars€am€aki   
1  03_article_04.pdf.txt  0.079849  0.387795   berenice         egypt   
2  03_article_01.pdf.txt  0.036107  0.337456         sv        verona   
3  03_article_05.pdf.txt  0.065030  0.353009  australia      plos one   
4  03_article_03.pdf.txt  0.020322  0.363023  lithuania            ad   
5  03_article_02.pdf.txt  0.105157  0.396240    driesch    kratochvil   

           5  
0      turku  
1     figure  
2      italy  
3       spit  
4     viking  
5  boessneck

**H2**

Our dataframe needs better column names. Put the appropriate values into the list below on line 1 to update the column names.

In [ ]:
column_names = ["article_file","polarity","subjectivity","keyword_one","keyword_two","keyword_three"]
pdf_dataframe.columns = column_names
pdf_dataframe

article_file  polarity  subjectivity keyword_one   keyword_two  \
0  03_article_00.pdf.txt  0.025864      0.355974     finland  k€ars€am€aki   
1  03_article_04.pdf.txt  0.079849      0.387795    berenice         egypt   
2  03_article_01.pdf.txt  0.036107      0.337456          sv        verona   
3  03_article_05.pdf.txt  0.065030      0.353009   australia      plos one   
4  03_article_03.pdf.txt  0.020322      0.363023   lithuania            ad   
5  03_article_02.pdf.txt  0.105157      0.396240     driesch    kratochvil   

  keyword_three  
0         turku  
1        figure  
2         italy  
3          spit  
4        viking  
5     boessneck

# Save our data

We'll save our dataframe as a CSV file and put it in our usual folder.

In [ ]:
pdf_dataframe.to_csv("adv_week3_homework.csv", index=False)
!cp adv_week3_homework.csv /content/gdrive/MyDrive/LibraryJuicePython

**H3** Take a moment to reflect on what your data analysis says about your PDF files. Jot down some reflections in the cell below.

My biggest takeaway here is that the text analysis for my articles--especially in so few keywords--is pretty flat. Most of the words come from the geographic locations of the sites; one article has the publication name as a keyword (from the headers and footers of the pdf).

I'd be very curious to do an analysis based on only the abstract, introduction, and conclusion of each article and see how the results compare. My hypothesis is that it would be more illuminating as to the subject matter of the articles, but I'd love to see those results. As it stands, the keyword analysis gives no indication that these articles are about animal burial sites.

Polarity values all fall between .02 and .11, and subjectivity values between .33 and .4


# All Done!

Congrats you are finished with this week's work. Don't forget to click the `Share` button above and enter `libraryjuicepresspython@gmail.com` so I can get a shared copy of your work. Save this as a PDF and upload it to the Library Juice Academy site.